### PR TITLE
Add wrappers for scrypt key derivation and password hashing

### PR DESCRIPTION
**Pull request by hsivonen**
Closes [issue 21](https://github.com/stef/pysodium/issues/21).

### CODE REVIEW

**Comment by stef** [comment]
awesome! please add testcases for all new functions to test/test_pysodium.py and if they pass i'll merge this.

thanks a lot!

File: test/test_pysodium.py [hunk 1]
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
         self.assertEqual(pysodium.crypto_generichash_blake2b_salt_personal(message, salt     = key[0:8]), binascii.unhexlify(b'11c29bf7b91b8500a463f27e215dc83afdb71ed5e959f0847e339769c4835fc7'))
         self.assertEqual(pysodium.crypto_generichash_blake2b_salt_personal(message, personal = key, key = key), binascii.unhexlify(b'5a0b3db4bf2dab71485211447fc2014391228cc6c1acd2f3031050a9a32ca407'))
 
+    def test_crypto_pwhash_scryptsalsa208sha256(self):
+        passwd =       b'Correct Horse Battery Staple'
```
Comment (bgaifullin) [inline test/test_pysodium.py]: Multiple spaces after '=', please fix.

File: test/test_pysodium.py [hunk 2]
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
...
+
+        key = pysodium.crypto_pwhash_scryptsalsa208sha256(32, passwd, salt, ops_limit, mem_limit)
+        self.assertEqual(len(key), 32)
+        self.assertEqual(key, binascii.unhexlify(b'34f05e9bef8beccd658acf5f123680b7d30c88d7e9328f9e47ab90185b6ee9ff'))
+
+        self.assertNotEqual(key, pysodium.crypto_pwhash_scryptsalsa208sha256(32, passwd, other_salt, ops_limit, mem_limit))
+        self.assertNotEqual(key, pysodium.crypto_pwhash_scryptsalsa208sha256(32, other_passwd, salt, ops_limit, mem_limit))
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str(self):
+        passwd =       b'Correct Horse Battery Staple'
```
Comment (bgaifullin) [inline test/test_pysodium.py]: multiple spaces after '='

File: test/test_pysodium.py [hunk 3]
```diff
@@ -121,5 +121,55 @@ def test_crypto_blake2b(self):
...
+        ops_limit = 1024
+
+        storage_string = pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit)
+        self.assertTrue(storage_string.startswith(pysodium.crypto_pwhash_scryptsalsa208sha256_STRPREFIX))
+        self.assertFalse(b'\x00' in storage_string)
+
+        self.assertNotEqual(storage_string, pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit), "Each call should compute a new random salt.")
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str_verify(self):
+        passwd =       b'Correct Horse Battery Staple'
```
Comment (bgaifullin) [inline test/test_pysodium.py]: multiple spaces after '='

**Comment by hsivonen** [comment]
@bgaifullin:

I've removed the multiple spaces even if it makes it less obvious visually what the different between `passwd` and `other_passwd` and `salt` and `other_salt` is.

@stef:
> awesome! please add testcases for all new functions to test/test_pysodium.py and if they pass i'll merge this.

It looks like instead of merging this PR, you merged [pull request 20](https://github.com/stef/pysodium/pull/20) instead even though the code there is incorrect on multiple counts. :-(

Should I expect PR 20 to be backed out and this PR to be merged in the current form of this PR or should I expect to have to reformulate this PR as a fix on top of that one?

**Comment by hsivonen** [comment]
> Should I expect PR 20 to be backed out and this PR to be merged in the current form of this PR or should I expect to have to reformulate this PR as a fix on top of that one?

I went ahead and did the latter.

**Comment by hsivonen** [comment]
Note that I intentionally did not provide default values for `memlimit` and `opslimit` to force the caller to think about them as the caller should.

File: test/test_pysodium.py [hunk 4]
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
...
+        self.assertEqual(key, binascii.unhexlify(b'34f05e9bef8beccd658acf5f123680b7d30c88d7e9328f9e47ab90185b6ee9ff'))
+
+        self.assertNotEqual(key, pysodium.crypto_pwhash_scryptsalsa208sha256(32, passwd, other_salt, ops_limit, mem_limit))
+        self.assertNotEqual(key, pysodium.crypto_pwhash_scryptsalsa208sha256(32, other_passwd, salt, ops_limit, mem_limit))
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str(self):
+        passwd = b'Correct Horse Battery Staple'
+
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
Comment (bgaifullin) [inline test/test_pysodium.py]: space between *

File: test/test_pysodium.py [hunk 5]
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
...
+        ops_limit = 1024
+
+        storage_string = pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit)
+        self.assertTrue(storage_string.startswith(pysodium.crypto_pwhash_scryptsalsa208sha256_STRPREFIX))
+        self.assertFalse(b'\x00' in storage_string)
```
Comment (bgaifullin) [inline test/test_pysodium.py]: self.assertIn

File: test/test_pysodium.py [hunk 6]
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
...
+
+        self.assertNotEqual(storage_string, pysodium.crypto_pwhash_scryptsalsa208sha256_str(passwd, ops_limit, mem_limit), "Each call should compute a new random salt.")
+
+    def test_crypto_pwhash_scryptsalsa208sha256_str_verify(self):
+        passwd = b'Correct Horse Battery Staple'
+        other_passwd = b'correct horse battery staple'
+        
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
Comment (bgaifullin) [inline test/test_pysodium.py]: space around *

File: test/test_pysodium.py [hunk 7]
```diff
@@ -121,18 +121,55 @@ def test_crypto_blake2b(self):
...
-        with self.assertRaises(ValueError):
-            pysodium.crypto_pwhash_scryptsalsa208sha256_str_verify(pwhash, 'wrongpassword')
+    def test_crypto_pwhash_scryptsalsa208sha256(self):
+        passwd = b'Correct Horse Battery Staple'
+        other_passwd = b'correct horse battery staple'
+        salt = binascii.unhexlify(b'4206baae5578933d7cfb315b1c257cc7af162965a91a74ccbb1cfa1d747eb691')
+        other_salt = binascii.unhexlify(b'4206baae5578933d7cfb315b1c257cc7af162965a91a74ccbb1cfa1d747eb692')
+        
+        # Use very small limits to avoid burning resources in CI
+        mem_limit = 32*1024
```
Comment (bgaifullin) [inline test/test_pysodium.py]: space around

**Comment by hsivonen** [comment]
@stef, could you, please, clarify what the expectations are here?

I have added tests and I have addressed @bgaifullin's review comments. The code itself has been correct all along. Yet, instead of this PR getting merged, first incorrect code gets merged from [PR 20](https://github.com/stef/pysodium/pull/20) and now instead of merging this PR, which would fix multiple bugs introduced in PR 20, [PR 25](https://github.com/stef/pysodium/pull/25) got priority even though it only fixed one problem with PR 20 and even the fix is bad (`.raw[:-1]` instead of `.value').

The code introduced in PR 20 continues to be dangerously incorrect in its salt handling, fails to take the output length in the first function from the caller and lacks the `crypto_pwhash_scryptsalsa208sha256_STRPREFIX` constant. The proposed changes here have been correct since before PR 20 was merged.

**Comment by stef** [comment]
thanks for the fixes

**Comment by hsivonen** [comment]
Thank you for the merge.

**Pull request merged**